XenAdmin/Dialogs/AcceptEuaDialog.cs

```diff
@@ -0,0 +1,126 @@
+﻿/* Copyright (c) Cloud Software Group, Inc. 
+ * 
+ * Redistribution and use in source and binary forms, 
+ * with or without modification, are permitted provided 
+ * that the following conditions are met: 
+ * 
+ * *   Redistributions of source code must retain the above 
+ *     copyright notice, this list of conditions and the 
+ *     following disclaimer. 
+ * *   Redistributions in binary form must reproduce the above 
+ *     copyright notice, this list of conditions and the 
+ *     following disclaimer in the documentation and/or other 
+ *     materials provided with the distribution. 
+ * 
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND 
+ * CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, 
+ * INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF 
+ * MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE 
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR 
+ * CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, 
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, 
+ * BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR 
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS 
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, 
+ * WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING 
+ * NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE 
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF 
+ * SUCH DAMAGE.
+ */
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+using XenAdmin.Core;
+using XenAPI;
+
+
+namespace XenAdmin.Dialogs
+{
+    public partial class AcceptEuaDialog : XenDialogBase
+    {
+        private readonly List<Host> _hosts;
+        private readonly Uri _targetUri;
+        private readonly HashSet<string> _euas;
+        private readonly HashSet<string> _errors;
+        private readonly HashSet<string> _warnings;
+
+
+        public AcceptEuaDialog(List<Host> hosts, Uri targetUri)
+        {
+            InitializeComponent();
+            _hosts = hosts;
+            _targetUri = targetUri;
+            _euas = new HashSet<string>();
+            _errors = new HashSet<string>();
+            _warnings = new HashSet<string>();
+
+            SetLoading(true);
+            warningTableLayoutPanel.Visible = false;
+
+            ThreadPool.QueueUserWorkItem(LoadEuas, null);
+        }
+
+        private void LoadEuas(object _)
+        {
+           _hosts.AsParallel().ForAll(FetchHostEua);
+
+            Program.BeginInvoke(this, () =>
+            {
+                if (_euas.Count > 1)

```

Comment:
We don't need to and they should all show the same. But the idea is that I'm adding a fairly small amount of logic to support this in the future. While if we don't, we might forget that XenCenter only supports one EUA per RPU. I think in cases like these the extra complexity is warranted since it does prevent issues in case things change.

Reply:
The problem with this though is that if the customer upgrades 5 pools, they will see a long text containing 5 times the same content, which is a bug from the user's perspective.

Reply:
I see what you mean but they're stored in a `HashSet` so it's only going to show more than one if they're different.

Reply:
Ah true, I forgot it was a HashSet! 🤦 Still it leaves the issue of fetching 5 times the same EUA which means extra delay.